lib/archethic/contracts.ex

```diff
@@ -59,29 +57,88 @@ defmodule Archethic.Contracts do
   Execute the contract trigger.
   """
   @spec execute_trigger(
-          Contract.trigger_type(),
-          Contract.t(),
-          nil | Transaction.t(),
-          nil | Recipient.t(),
-          Keyword.t()
-        ) :: {:ok, nil | Transaction.t()} | {:error, String.t()}
-  defdelegate execute_trigger(
-                trigger_type,
-                contract,
-                maybe_trigger_tx,
-                maybe_recipient,
-                opts \\ []
-              ),
-              to: Interpreter,
-              as: :execute_trigger
+          trigger :: Contract.trigger_type(),
+          contract :: Contract.t(),
+          maybe_trigger_tx :: nil | Transaction.t(),
+          maybe_recipient :: nil | Recipient.t(),
+          maybe_state_utxo :: nil | UnspentOutput.t(),
+          opts :: Keyword.t()
+        ) :: Failure.t() | ActionWithTransaction.t() | ActionWithoutTransaction.t()
+  def execute_trigger(
+        trigger_type,
+        contract = %Contract{transaction: contract_tx},
+        maybe_trigger_tx,
+        maybe_recipient,
+        maybe_state_utxo \\ nil,
+        opts \\ []
+      ) do
+    state = State.from_utxo(maybe_state_utxo)
+
+    # TODO: trigger_tx & recipient should be transformed into recipient here
+    # TODO: rescue should be done in here as well
+    # TODO: implement timeout
+
+    case Interpreter.execute_trigger(
+           trigger_type,
+           contract,
+           state,
+           maybe_trigger_tx,
+           maybe_recipient,
+           opts
+         ) do
+      {:ok, nil, next_state, logs} ->
+        # I hate you credo
+        execute_trigger_noop_response(
+          State.to_utxo(next_state),
+          logs,
+          maybe_state_utxo,
+          contract_tx
+        )
+
+      {:ok, next_tx, next_state, logs} ->
+        case State.to_utxo(next_state) do
+          {:ok, maybe_utxo} ->
+            %ActionWithTransaction{
+              logs: logs,
+              next_tx: next_tx,
+              next_state_utxo: maybe_utxo
```

Comment:
I think it would be easier to return the encoded state directly instead of an UTXO. Actually this state is used in some places after that don't need the UTXO (Fee calculation for example).
The utxo casting could be done only in the consume_inputs of LedgerOperation maybe